v1.7 backports 2020-08-04

rolinh opened the pull request:
* #12729 -- datapath: Fix ICMP ECHO tuple ports (@brb)
 * #12409 -- Cilium operator HA mode (@fristonio)
 * #12771 -- install/kubernetes: do not schedule cilium-operator pods in same node (@aanm)

Partially backported:
* https://github.com/cilium/cilium/pull/10997

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 12729 12409 12771; do contrib/backporting/set-labels.py $pr done 1.7; done
```

tklauser committed:
rand: rename RandomRune* funcs to RandomString*

[ upstream commit 3c9c9709cf6a82da9060fa824119c24463e18c0a ]

These functions return strings, not a rune. Rename them accordingly.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>
Signed-off-by: André Martins <andre@cilium.io>

fristonio committed:
vendor: vendor kubernetes leaderelection library

[ upstream commit 66c3d9c4d3ef85d57914ec1a595f0226fcde7e00 ]

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>
Signed-off-by: André Martins <andre@cilium.io>

fristonio committed:
k8s: add coordinationv1 capability check to k8s version package

[ upstream commit fb101dfc04ddb6277413207eb6b6580f4be82b82 ]

* Introduces config option `K8sLeasesFallbackDiscoveryEnabled` to check
  if fallback discovery is enabled for Leases.
* K8sLeasesFallbackDiscovery is enabled by default only in
  operator.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>
Signed-off-by: André Martins <andre@cilium.io>

fristonio committed:
operator: support HA mode for operator using k8s leaderelection library

[ upstream commit df90c99905ad107710ce66d2dd36820f068db189 ]

* Make leaderelection parameters configurable using command line flags
* Update cmdref to include documentation for new flags.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>
Signed-off-by: André Martins <andre@cilium.io>

fristonio committed:
install: update helm templates to add HA capabilities for operator

[ upstream commit 930bde726974320196583cde03ccf1c57af55606 ]

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>
Signed-off-by: André Martins <andre@cilium.io>

fristonio committed:
test: generate cilium helm template validating against k8s cluster

[ upstream commit 82cc7c3d076149e147cd120f7ee4424317a39ccd ]

* Use --validate with `helm template` command to validate the generated
  manifest against the associated kubernetes cluster
* For more information see -
  https://github.com/cilium/cilium/pull/12409#discussion_r453313631

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>
Signed-off-by: André Martins <andre@cilium.io>

brb committed:
datapath: Fix CT tuple ports for ICMP Echo

[ upstream commit c633c2d02d59cc9429800bd1912187f246eaf0ac ]

Previously, when an ICMP EchoRequest was sent from one node A to another
node B with Echo ID > NAT_MIN_EGRESS, the ICMP EchoReply sent from B -> A
created a CT entry and NAT entries which could not be related by GC.
E.g. node A (192.168.34.12) pings node B (192.168.34.11):

    ICMP IN 192.168.34.12:0 -> 192.168.34.11:38193 XLATE_DST
    192.168.34.11:38193 Created=6292sec HostLocal=1
    ICMP OUT 192.168.34.11:38193 -> 192.168.34.12:0 XLATE_SRC
    192.168.34.11:38193 Created=6292sec HostLocal=1

    ICMP OUT 192.168.34.11:0 -> 192.168.34.12:38193 expires=16783063
    RxPackets=0 RxBytes=0 RxFlagsSeen=0x00 LastRxReport=0 TxPackets=1
    TxBytes=50 TxFlagsSeen=0x00 LastTxReport=16783005 Flags=0x0000 [ ]
    RevNAT=0 SourceSecurityID=0 IfIndex=0

This made the NAT entries to escape the CT GC meaning that the CT entry
was removed, while the NAT entries were kept which made them to stay
forever until a user manually ran "cilium bpf nat flush".

Fix this by setting ICMP Echo ID in a port which belongs to addr of
the local node, so that the CT GC could relate the NAT entries. In the
previous example, the CT entry after the fix is the following:

    ICMP OUT 192.168.34.11:38193 -> 192.168.34.12:0 expires=16783063
    RxPackets=0 RxBytes=0 RxFlagsSeen=0x00 LastRxReport=0 TxPackets=1
    TxBytes=50 TxFlagsSeen=0x00 LastTxReport=16783005 Flags=0x0000 [ ]
    RevNAT=0 SourceSecurityID=0 IfIndex=0

The fix does not change the ID placement in a port for the case when
B -> A sends ICMP EchoRequest.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

brb committed:
ctmap: Add unit test for ICMP CT/NAT GC

[ upstream commit ae3917b64dd04d35c68e5746b1b670555a9f0ffe ]

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

brb reviewed (approved):
LGTM, thanks.

aanm committed:
install/kubernetes: do not schedule cilium-operator pods in same node

[ upstream commit bde2daf77fa8bac84a66c803ecc18f78df779082 ]

Since Cilium Operator is running in host network, 2 or more pods can't
run in the same node at the same time or they will clash the open
ports they use for liveness and / or readiness health check.

Fixes: 930bde726974 ("install: update helm templates to add HA capabilities for operator")
Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -16,6 +16,9 @@ config:
   enabled: true
 
+  # Number of replicas to run for cilium-operator deployment.
+  numReplicas: 2
```
Should we default this to 1 but allow override, given how late in the cycle this is for v1.7? It may be surprising to users that suddenly there are two cilium-operators in their cluster.

joestringer committed:
install: Default cilium-operator replicas to 1

Given that Cilium v1.7.x was not released with operator HA and we are
otherwise disabling this functionality (including etcd heartbeats) by
default in v1.7 branch, we should also not surprise users during the
v1.7.7->v1.7.8 point release update by upgrading to find that there are
more copies of the cilium-operator running in their cluster.

Users can still opt in by explicitly configuring the number of replicas.

Signed-off-by: Joe Stringer <joe@cilium.io>
